### ISSUE TITLE
deriving from `Jason.Encoder` requires you to add Jason as a boundary.

### ISSUE THREAD

**Issue by tcoopman** [sasa1977/boundaries#3]
Opening for discussion. It's not really an issue, in that you have to add this line, but it's also not the cleanest.
Maybe someone has an idea about what to do with this?

**Comment by sasa1977** [comment]
I need to try to repro later (feel free to post a minimal version of the code if you have it), but that shouldn't be happening at all, since Jason is an external dependency.

**Comment by tcoopman** [comment]
I don't have a minimal example, but do have an example of the error:

```bash
warning: Jason.Encoder.BizRules.Message.Event.RuleWasEnabled is not included in any boundary
  lib/biz_rules/message/event/rule_was_enabled.ex
```

**Comment by sasa1977** [comment]
That sounds like your own custom module, not the Jason module, right? If so, it is a part of your codebase, and thus it needs to be included in some boundary.

**Comment by tcoopman** [comment]
I've created a minimal reproduction: https://github.com/tcoopman/boundaries-bug-3

```bash
> mix compile

warning: Jason.Encoder.Foo.Bar is not included in any boundary
  lib/foo/bar.ex

```

**Comment by sasa1977** [comment]
OH, I see. OK, that sucks :( It seems that the problem would appear for any `@derive`.

Ideally, `boundaries` would detect this situation and autoexclude the module. Not sure if that's possible though.

If that doesn't work, then we'll have to add some mechanism for removing modules from being checked, or otherwise drop the "not included in any boundary" warning completely.

**Comment by sasa1977** [comment]
OK, I've briefly played with this, and my finding is that this problem is related to all protocol implementations, derived or underived. 

Consider the following module:

```elixir
defmodule Test do
  @derive [Jason.Encoder]
  defstruct x: 1, y: 2

  defimpl String.Chars do
    def to_string(_x), do: "#Test"
  end
end
```

This will generate two impl modules: `Jason.Encoder.Test` and `String.Chars.Test`. Both modules belong to the project, but their namespaces are "skewed", and now the user must declare boundaries for `Jason.Encoder.Test` and `String.Chars.Test`. This can become cumbersome, especially in projects where many modules implement many protocols.

We could detect this situation by looking at the undocumented `:protocol_impl` attribute:

```elixir
iex> String.Chars.Test.__info__(:attributes)[:protocol_impl]
[protocol: String.Chars, for: Test]
```

We could proclaim that the implementer module (`String.Chars.Test`) falls under the same boundary as the implementee (`Test`), but the question is should we?

I'm not completely convinced. The thing about protocols is that anyone can provide an implementation. So e.g. you could have a UI layer implementing a protocol for some domain structure. In such case, we shouldn't bound the implementation to the same rules as the domain structure, because the implementation is really a UI concern. In other words, even though a `defimpl` is transforming the domain struct, it might need to follow the UI boundary rules.

So, at this point, I'm not completely sure how to proceed. Ignoring is certainly an option, but I don't think it's the best one, because custom defimpls are still code which should be verified.

One approach that comes to mind is the following:

- By default, a protocol impl is ignored.
- The writer of `defimpl` can specify the target boundary of the protocol. This would probably move us into the direction of decentralized configuration, as proposed [here](https://elixirforum.com/t/boundaries-enforcing-boundaries-in-elixir-projects/24623/16?u=sasajuric).

**Comment by sasa1977** [comment]
I also addressed this in #5, which accommodates protocol impls. The unclassified modules support should address `@derive`, although I feel this could be even done automatically, without requiring you to add anything to the ignore list.
